drivers/interrupt_controller/intc_esp32c3.c

```diff
@@ -35,129 +33,116 @@ LOG_MODULE_REGISTER(intc_esp32c3, CONFIG_LOG_DEFAULT_LEVEL);
-#define ESP32C3_INTC_SRCS_PER_IRQ       2
-#define ESP32C3_INTC_AVAILABLE_IRQS     30
+#define ESP32C3_INTC_DEFAULT_PRIORITY  15
+#define ESP32C3_INTC_DEFAULT_THRESHOLD 1
+#define ESP32C3_INTC_DISABLED_SLOT     31
+#define ESP32C3_INTC_SRCS_PER_IRQ      2
 
+/* Define maximum interrupt sources per SoC */
 #if defined(CONFIG_SOC_SERIES_ESP32C6)
-
-#define IRQ_NA		0xFF	/* IRQ not available */
-#define IRQ_FREE	0xFE
-
-#define ESP32C6_INTC_SRCS_PER_IRQ       2
-#define ESP32C6_INTC_AVAILABLE_IRQS     31
-
-/* Interrupt overview for ESP32C6:
- * - 0, 3, 4, and 7 are used by the CPU for core-local interrupts (CLINT)
- * - 1 is used for Wi-Fi in Espressif HAL
- * - 2, 5, 6, 8 .. 31 are available for Zephyr
- * - 31 is reserved for disabled interrupts
+/*
+ * Interrupt reserved mask
+ * 0 is reserved
+ * 1 is for Wi-Fi
+ * 3, 4 and 7 are unavailable for PULP CPU as they are bound to Core-Local Interrupts (CLINT)
+ */
+#define rsvd_mask (BIT(0) | BIT(1) | BIT(3) | BIT(4) | BIT(7))
+#define ESP_INTC_AVAILABLE_IRQS 31
+#else /* CONFIG_SOC_SERIES_ESP32C2 && CONFIG_SOC_SERIES_ESP32C3 */
+/*
+ * Interrupt reserved mask
+ * 0 is reserved
+ * 1 is for Wi-Fi
  */
-static uint8_t esp_intr_irq_alloc[ESP32C6_INTC_AVAILABLE_IRQS][ESP32C6_INTC_SRCS_PER_IRQ] = {
-	[0] = {IRQ_NA, IRQ_NA},
-	[1] = {IRQ_NA, IRQ_NA},
-	[2] = {IRQ_FREE, IRQ_FREE},
-	[3] = {IRQ_NA, IRQ_NA},
-	[4] = {IRQ_NA, IRQ_NA},
-	[5 ... 6] = {IRQ_FREE, IRQ_FREE},
-	[7] = {IRQ_NA, IRQ_NA},
-	[8 ... 30] = {IRQ_FREE, IRQ_FREE}
-};
+#define rsvd_mask (BIT(0) | BIT(1))
+#define ESP_INTC_AVAILABLE_IRQS 30
 #endif
 
-#define STATUS_MASK_NUM		3
+/* Single array for IRQ allocation */
+static uint8_t esp_intr_irq_alloc[ESP_INTC_AVAILABLE_IRQS * ESP32C3_INTC_SRCS_PER_IRQ];
 
-static uint32_t esp_intr_enabled_mask[STATUS_MASK_NUM] = {0, 0, 0};
+#define ESP_INTR_IDX(irq, slot) ((irq % ESP_INTC_AVAILABLE_IRQS) * ESP32C3_INTC_SRCS_PER_IRQ + slot)
 
-#if defined(CONFIG_SOC_SERIES_ESP32C2) || defined(CONFIG_SOC_SERIES_ESP32C3)
+#define STATUS_MASK_NUM 3
+
+static uint32_t esp_intr_enabled_mask[STATUS_MASK_NUM] = {0, 0, 0};
 
 static uint32_t esp_intr_find_irq_for_source(uint32_t source)
 {
-	/* in general case, each 2 sources goes routed to
-	 * 1 IRQ line.
-	 */
-	uint32_t irq = (source / ESP32C3_INTC_SRCS_PER_IRQ);
-
-	if (irq > ESP32C3_INTC_AVAILABLE_IRQS) {
-		INTC_LOG("Clamping the source: %d no more IRQs available", source);
-		irq = ESP32C3_INTC_AVAILABLE_IRQS;
-	} else if (irq == 0) {
-		irq = 1;
+	if (source >= ETS_MAX_INTR_SOURCE) {
+		return IRQ_NA;
 	}
 
-	INTC_LOG("Found IRQ: %d for source: %d", irq, source);
+	uint32_t irq = source / ESP32C3_INTC_SRCS_PER_IRQ;
 
-	return irq;
-}
+	/* Check if the derived IRQ is usable first */
+	for (int j = 0; j < ESP32C3_INTC_SRCS_PER_IRQ; j++) {
+		int idx = ESP_INTR_IDX(irq, j);
 
-#elif defined(CONFIG_SOC_SERIES_ESP32C6)
+		/* Ensure idx is within a valid range */
+		if (idx >= (ESP_INTC_AVAILABLE_IRQS * ESP32C3_INTC_SRCS_PER_IRQ)) {
+			continue;
+		}
```

Comment:
I suggest to use `ARRAY_SIZE` just in case the array is declared differently in the future.

```suggestion
		/* Ensure idx is within a valid range */
		if (idx >= ARRAY_SIZE(esp_intr_irq_alloc)) {
			continue;
		}
```